app/uk/gov/hmrc/hec/services/scheduleService/HecTaxCheckExtractionService.scala

```diff
@@ -42,57 +44,122 @@ class HecTaxCheckExtractionServiceImpl @Inject() (
+          )
         _                <- createAndStoreFile(
                               LicenceValidityPeriod,
                               seqNum,
                               licenceValidityPeriod.partialFileName,
-                              licenceValidityPeriod.dirName
+                              licenceValidityPeriod.dirName,
+                              false
                             )
-        //updating isExtracted to true for the the processed hec tax check codes
-        newHecTaxCheck   <- taxCheckService.updateAllHecTaxCheck(updatedHecTaxChek)
-      } yield newHecTaxCheck
+        _                <- createAndStoreFile(
+                              CorrectiveAction,
+                              seqNum,
+                              correctiveAction.partialFileName,
+                              correctiveAction.dirName,
+                              false
+                            )
+        hecTaxCheck      <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false)
+        _                <- createHecFile(HECTaxCheckFileBodyList(hecTaxCheck), count, hecData.partialFileName, hecData.dirName)
+        updatedHecTaxChek = hecTaxCheck.map(_.copy(isExtracted = true))
+
+        //i think we don't need to bring the whole list of codes which are updated ,
+        // it may cause performance issue, so now it's returning Unit.
+        _ <- taxCheckService.updateAllHecTaxCheck(updatedHecTaxChek)
+      } yield ()
     }
     result.value
   }
 
+  private def createHecFile(
+    hecTaxCheckList: HECTaxCheckFileBodyList,
+    count: Int,
+    partialFileName: String,
+    dirname: String
+  ) = {
+    @tailrec
+    def loopHecTaxCheckRecords(
+      hecTaxCheckList: List[HECTaxCheck],
+      count: Int,
+      seqNumInt: Int,
+      partialFileName: String,
+      dirname: String
+    ): EitherT[Future, Error, Unit] = {
+      val hecList       = hecTaxCheckList.take(count)
+      val remainingList = hecTaxCheckList.drop(count)
+      val isRemaining   = remainingList.size > 0
+      //If list is empty or sequence number exceeds 19999, then halt the process
+      if (hecList.size === 0 || seqNumInt > 19999) {
+        logger.info(
+          "Hec tax Check file creation process halted:: Either there are no more records to process or the sequence number exceeds 9999. "
+        )
+        EitherT.fromEither[Future](Right(()))
+      } else {
+        val _ = createAndStoreFile(
+          HECTaxCheckFileBodyList(hecList),
+          seqNumInt.toString.takeRight(4),
+          partialFileName,
+          dirname,
+          isRemaining
+        )
+        loopHecTaxCheckRecords(remainingList, count, seqNumInt + 1, partialFileName, dirname)
+      }
+    }
+
+    //Started with 10001, so that we can do increment and then takeRight(4), will become the sequence number.
+    //This done as addition t 0001 will give just 2
+    loopHecTaxCheckRecords(hecTaxCheckList.list, count, 10001, partialFileName, dirname)
```

Comment:
if we wanted something that didn't look as funny we could start at 1 here and then do something like this to pad with zeros  to the correct length when we need the sequence number string:
```scala
def f(i: Int): String = f"$i%04d"
f(1) // = "0001"
```